Migration of zuliprc location to a config directory

Niloth-p opened the pull request:
### What does this PR do, and why?

Adds the following features:

- Support recognizing zuliprc from multiple select locations. 
  - `~/zuliprc`
  - the user's downloads directory
  - the user's config directory

- Move zuliprc creation into user's config directory.
The new path would look like `.config/zulip-terminal/zuliprc`.

- Create realm-wise zuliprc files in the configuration directory.
Each realm gets its own directory.
The new directory structure would look like:
```
<user's config directory>/zulip-terminal/
  org1-name/zuliprc
  org2-name/zuliprc
  zuliprc
```

- Add new positional argument - prefix of realm name to get its configuration.
New command examples:
`zulip-term zuli`
`zulip-term "full organization name"`

- Add -o argument to list realms with stored zuliprc.
```
$zulip-term -o
The following organizations are available:
  Zulip Community
  Another Organization
```

- Add -n argument to login to a realm, and create config.
```
$zulip-term -n
Please enter your credentials to login into your Zulip organization.
```

<details>
<summary>What does this PR do, and why? (expired)</summary>

### What does this PR do, and why?
Moves config location to 'XDG_CONFIG/zulip/'.
Makes zuliprc files dot files - .zuliprc.
Use independent org.zuliprc files for each organization.

Does not yet separate the zterm config from the api config.
Does not yet update the documents (README.md, FAQ.md, etc).
New tests not added.

The code is not yet meant for review.
This is only a draft to play around with the configuration in order to finalize what behavior is desired.
</details>

### External discussion & connections
- [x] Discussed in **#zulip-terminal** in `Moving config`
- [ ] Fully fixes #
- [x] Partially fixes issue #678
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
- [x] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

Niloth-p commented on zulipterminal/cli/run.py:
```diff
@@ -49,7 +49,7 @@ class SettingData(NamedTuple):
 
 DOWNLOADED_PATH_ZULIPRC = str(downloads_file_path() / "zuliprc")
 HOME_PATH_ZULIPRC = str(Path.home() / "zuliprc")
-CONFIG_PATH_ZULIPRC = str(config_file_path() / "zulip-terminal" / "zuliprc")
```
Changed names here.
Felt CONFIG_PATH_ZULIPRC was a more consistent name for the single zuliprc in config.
Feel ZULIP_CONFIG_PATH would be better when pointing to the directory. Could use CONFIG_PATH_ZULIP too.
Would be open to rebasing any name changes if the noise of this change is more important than the consistency of the name in the initial commits.

Niloth-p commented on zulipterminal/cli/run.py:
```diff
@@ -343,13 +349,19 @@ def _write_zuliprc(
+    valid_locations.extend(
+        str(file) for file in Path(ZULIP_CONFIG_PATH).glob("zuliprc")
+    )
+    valid_locations.extend(
```
The current implementation explicitly does not use `Path(ZULIP_CONFIG_PATH).glob("**/zuliprc")`, to avoid the situation (in later commits) where the realm has multiple zuliprc files inside it (in folders or as loose file). This implementation will only recognize the outermost zuliprc within a realm's directory, irrespective of the subdirectories' contents.

Niloth-p commented on zulipterminal/cli/run.py:
```diff
@@ -448,6 +448,30 @@ def list_themes() -> str:
+        exit_with_error(
+            f"Could not find any organizations starting with '{realm_name_prefix}'"
+        )
+    matching_files = list(matching_dirs[0].glob("zuliprc"))
```
If we used path.glob(**/zuliprc) in the previous commit, this is where we would need to add another case where len(matching_files) > 1.

Niloth-p commented on zulipterminal/cli/run.py:
```diff
@@ -448,6 +448,30 @@ def list_themes() -> str:
 
 
+def path_to_realm_zuliprc(realm_name_prefix: str) -> str:
+    if len(realm_name_prefix) < 3:
```
Assumption: There are no zulip realms whose name property contains 2 or fewer characters.
We could remove this case.

Niloth-p commented on zulipterminal/cli/run.py:
```diff
@@ -387,15 +398,21 @@ def check_for_default_zuliprc() -> str:
     while zuliprc_path == "" or not path.exists(zuliprc_path):
         try:
-            zuliprc_path = fetch_zuliprc(zuliprc_path)
+            zuliprc_path = fetch_zuliprc(zuliprc_path, new_realm)
```
Re. naming,
used 'realm' for all internal references,
and 'org' / 'organization' for all references visible to the user.

Niloth-p reviewed (commented):
Updated the PR description, to describe the new features.
Each item in that feature list is its own commit (in the same order), with preparatory commits in between (mentioned for easier reviewing).

I have not yet added tests in this branch.
The ones I had were too tightly tied to the functions that were introduced.
Planning to re-write them to directly test main([]) instead.

#### Reference: Testing cases:
Feel free to come up with any missing / edge cases.
- default call (without positional argument)
	- loose zuliprc file in config
	- zuliprc files in config inside realms directories
		- one realm
		- multiple realms
	- mix of both loose zuliprc and inside realms in config
	- several zuliprc files found error
	- no zuliprc in any locations
- positional argument
	- too few characters (<3)
	- no org found
	- several matching orgs
	- one matching org, but zuliprc is inside subdirectory
	- one matching org
	- full name instead of prefix
	- use along with --config-file option & -c options
- login newly into a few servers, test the created directory structure
- n option:
	- if configuration for that realm already exists
	- if a directory with that realm name exists, but it does not have a zuliprc file inside it
- o option:
	- no orgs found
	- several orgs found

rsashank commented on zulipterminal/cli/run.py:
```diff
@@ -322,11 +371,48 @@ def _write_zuliprc(
-def parse_zuliprc(zuliprc_str: str) -> Dict[str, SettingData]:
-    zuliprc_path = path.expanduser(zuliprc_str)
-    while not path.exists(zuliprc_path):
+def check_for_default_zuliprc() -> str:
+    zuliprc_locations = [
+        DOWNLOADED_PATH_ZULIPRC,
+        HOME_PATH_ZULIPRC,
+    ]
+    valid_locations = [
+        location for location in zuliprc_locations if Path(location).exists()
+    ]
+    valid_locations.extend(
+        str(file) for file in Path(ZULIP_CONFIG_PATH).glob("zuliprc")
+    )
+    valid_locations.extend(
+        str(file) for file in Path(ZULIP_CONFIG_PATH).glob("*/zuliprc")
+    )
+
+    count = len(valid_locations)
+    if count > 1:
+        valid_locations_string = "\n".join(valid_locations)
+        exit_with_error(
+            f"Found multiple zuliprc files at:\n{valid_locations_string}\n"
+            "Please retry by specifying the path to your target zuliprc file."
+        )
+    elif count == 1:
+        return valid_locations[0]
+    return ""
```
This implementation seems great! However, should we consider giving priority to `HOME_PATH_ZULIPRC` if multiple configuration files are found, since that was previously the location of zuliprc?

If users want to use a different configuration file instead of the default, they can specify a specific target zuliprc file to override the default. We can notify the config file used while loading ZT, similar to how we do with the settings.

Alternatively, instead of exiting with an error, we could include an index and prompt the user to choose which organization they would like to log into.

For example:

```shell
Choose which organization you'd like to log in to (Enter the number):
1: Zulip Community: chat.zulip.org
2: CTF Chat: ctf.zulipchat.com
3: Example: example.zulip.com
```

rsashank commented on zulipterminal/cli/run.py:
```diff
@@ -392,6 +478,40 @@ def list_themes() -> str:
+def path_to_realm_zuliprc(realm_name_prefix: str) -> str:
+    if len(realm_name_prefix) < 3:
+        exit_with_error("Organization name prefix must be at least 3 characters long")
+    matching_dirs = [
+        dir
+        for dir in Path(ZULIP_CONFIG_PATH).glob("*/")
+        if dir.name.startswith(realm_name_prefix)
+    ]
```
Can we make this case sensitive?

Maybe:

```python
    matching_dirs = [
        dir
        for dir in Path(ZULIP_CONFIG_PATH).glob("*/")
        if dir.name.startswith(realm_name_prefix)
        or dir.name.lower().startswith(realm_name_prefix)
    ]
```

rsashank commented on zulipterminal/cli/run.py:
```diff
@@ -392,6 +478,40 @@ def list_themes() -> str:
+        for dir in Path(ZULIP_CONFIG_PATH).glob("*/")
+        if dir.name.startswith(realm_name_prefix)
+    ]
+    if len(matching_dirs) > 1:
+        matching_dirs_string = "\n".join(str(dir) for dir in matching_dirs)
+        exit_with_error(
+            f"Found multiple organizations starting with '{realm_name_prefix}':\n"
+            f"{matching_dirs_string}\n"
+        )
```
I'm not sure how feasible this is, but should we consider prompting the user to select which directory to log into from the matched directories?

rsashank commented on zulipterminal/cli/run.py:
```diff
@@ -392,6 +478,40 @@ def list_themes() -> str:
+            f"Found multiple organizations starting with '{realm_name_prefix}':\n"
+            f"{matching_dirs_string}\n"
+        )
+    elif len(matching_dirs) == 0:
+        exit_with_error(
+            f"Could not find any organizations starting with '{realm_name_prefix}'"
+        )
```
Maybe we could suggest using zulip-term -o to show the list of available organizations. Alternatively, we could display the list along with the error message?

rsashank reviewed (commented):
Great work, @Niloth-p! I manually tested it, and it works well. I added a few comments in some areas where I think it would be nice to consider changes, but those are more opinion-based.

This looks very useful!

rsashank commented:
Thanks for updating this, @Niloth-p! I'll leave it for the mentor and maintainers to review. I'm removing the buddy review label now.

Niloth-p committed:
platform_code: Add a function that gets the path to XDG_CONFIG_HOME.

neiljp commented on zulipterminal/cli/run.py:
```diff
@@ -452,6 +452,30 @@ def list_themes() -> str:
     )
 
 
+def path_to_realm_zuliprc(realm_name_prefix: str) -> str:
```
I'll wait to see this updated for the new version first, but I expect this and later commits would be the mapping of aliases to zuliprcs instead.

neiljp commented on zulipterminal/cli/run.py:
```diff
@@ -305,8 +311,13 @@ def fetch_zuliprc(zuliprc_path: str) -> str:
+    )
     print(
-        f"{in_color('red', f'zuliprc file was not found at {locations_checked}')}"
+        f"{missing_zuliprc_text}"
```
This will likely be simplified with the other changes I suggested?

neiljp commented on zulipterminal/cli/run.py:
```diff
@@ -157,6 +157,12 @@ def parse_args(argv: List[str]) -> argparse.Namespace:
         action="store_true",
         help="list all the organizations that you have configurations for, and exit",
     )
+    parser.add_argument(
+        "-n",
+        "--new-organization",
+        action="store_true",
+        help="login to a new organization",
+    )
```
It doesn't quite fit within the remit of the zuliprc movement work, but it'd be good to have `-c` work with creating a zuliprc file from username/password, which I don't think it does right now?

neiljp reviewed (commented):
@Niloth-p I hope this review helps get you back on track with this work :+1:

The discussion in the channel seems to have helped too :)

Niloth-p committed:
refactor: run: Split parse_zuliprc() to create a separate function.

Created resolve_to_valid_path() to return a valid zuliprc path.
Added docstring.

Niloth-p committed:
refactor: run: Use a constant for the default zuliprc path.

Niloth-p committed:
refactor: run: Rename fetch_zuliprc function, and add docstring.

Niloth-p committed:
refactor: run: Set zuliprc_path via login_and_save().

Previously, zuliprc_path was set directly in main().
Now, we are setting it 2 levels deeper, inside of login_and_save(),
passing it to main() through resolve_to_valid_path().

main() now sets the default value of zuliprc_path to None, and uses the
returned path from resolve_to_valid_path().

This is done to allow adding support for multiple zuliprc paths later,
where we will dynamically compute the path in login_and_save(), after
taking user input.

Niloth-p committed:
run: Use the new config directory as a default zuliprc path.

In addition to the previous default path ~/zuliprc.

Introduce the .config path of zulip-terminal as a str constant.

Add a function check_for_default_zuliprc() to search for zuliprc files
in both the default locations, when the --config-file argument is not
used.

It is added to ensure that users who have only a single account can
skip using the account alias argument that will be added.

Docstring updated.
Tests updated.

Niloth-p committed:
run: Create account-wise zuliprc files in the new config path.

Newly generated zuliprc files were being saved to `~/zuliprc`.
This moves default zuliprc creation into the user's config directory.

The --config-file option will no longer be able to save zuliprc files
in custom locations.

The new config path including the .config directory are handled safely,
to allow creation even if they do not already exist.

Tests updated.

The test with id: `path_does_not_exist`, in the test function
`test_main_cannot_write_zuliprc_given_good_credentials()` has been
removed, as the path will now be created even if it does not exist.

Niloth-p committed:
run: Add new positional argument `account-alias` to load config.

The usage of the account-alias positional argument is made mutually
exclusive with the config-file option.

Niloth-p committed:
run: Add new argument `--list-accounts`.

Lists all the accounts in the config path, by checking the directory
structure. The home path zuliprc is not considered in this list.

Tests updated.

Niloth-p committed:
run: List all accounts when the user enters an invalid account alias.

Niloth-p committed:
refactor: run: Stop passing zuliprc_path to login_and_save().

Previously, we had to pass the zuliprc_path because the --config-file
option could be used to create a zuliprc at a custom location.

Now, all accounts are only saved to the standard config path. So, by
moving the print statement to the outer function, we can simplify
login_and_save().

Niloth-p committed:
run: Add new argument `--new--account` to create a new account.

- If --new-account is used along with --config-file,
  - If the given path is valid, that zuliprc will be loaded
  - Else, a new account will be created, but not at the given path.
- If --new-account is used along with account-alias, the new account
argument will be ignored.

Tests updated.
